qbee-agent.service.in: Adjust order

leon-anavi opened the pull request:
Start qbee-agent.service.in after systemd-resolved.service and network-online.target to avoid the following warnings and errors:

```
[WARNING] failed to get config from API: failed to send API request: Get "https://device.app.qbee.io:443/v1/org/device/auth/config": dial tcp: lookup device.app.qbee.io on 1.0.0.1:53: dial udp 1.0.0.1:53: connect: network is unreachable
```

```
[ERROR] failed to do check-in: failed to send API request: Get "https://device.app.qbee.io:443/v1/org/device/auth/agent/arm64/checkin": dial tcp: lookup device.app.qbee.io on 1.0.0.1:53: dial udp 1.0.0.1:53: connect: network is unreachable
```

broale-dish commented:
I don't think it is correct to add systemd-resolved specifically, that couples the service to systemd. Even though "after" is lenient and this won't fail, I think network-online should be enough.

jonhenrik13 commented:
Thanks for the contribution. The fact that qbee-agent starts after the network target rather than the network-online target is deliberate. The agent will continue to work with local resources where it can (eg. setting of user passwords, ssh keys etc.) when network connectivity cannot be reached, and in general we want the agent to have as few external dependencies as possible. However, there's currently work being done making the work a bit more streamlined on initial start up. Essentially there will be a retry mechanism first time the agent is scheduled to run configuration (typically during boot). It will still log connectivity issues, but it will appear online sooner. I hope this would be good enough for your case.

broale-dish commented:
> > I don't think it is correct to add systemd-resolved specifically, that couples the service to systemd. Even though "after" is lenient and this won't fail, I think network-online should be enough.
> 
> `qbee-agent.service` is a systemd service so it has been already coupled to systemd.

My bad, I intended to say systemd-resolved, which is a optional subsystem of systemd. Not systemd itself.

jonhenrik13 commented:
Mender and qbee has different feature sets as qbee also aims to be a configuration management engine. That means that the qbee-agent could work solely with local resources even if there are no network connectivity (depending on how the agent is configured). What is meant by external dependencies in this case is not dependencies on local systemd services, but rather external network services like DHCP. My understanding is that if the device for some reason does not get a valid DHCP response, then it will never reach the network-online target and qbee-agent will not start.

The relevant work on the agent is in this PR: https://github.com/qbee-io/qbee-agent/pull/74

It's still subject to testing and review, but the gist of it is that it will do a retry for the initial api call to the qbee backend on startup.

leon-anavi commented:
> Mender and qbee has different feature sets as qbee also aims to be a configuration management engine. That means that the qbee-agent could work solely with local resources even if there are no network connectivity (depending on how the agent is configured). What is meant by external dependencies in this case is not dependencies on local systemd services, but rather external network services like DHCP. My understanding is that if the device for some reason does not get a valid DHCP response, then it will never reach the network-online target and qbee-agent will not start.

The reported error and warning from qbee-agent logs are always observed on the first boot with Raspberry Pi because of a systemd service race condition. I don't find it user friendly and I don't expect under normal circumstances to see errors and warning in the logs.

Eventually, just as an idea, if you want I can modify the GitHub pull request to have two separate systemd services: one that runs on the first boot and wants `network-online.target` (because the device has to register to qbee.io cloud) and another for next boots that runs just after `network.target`.

> The relevant work on the agent is in this PR: [qbee-io/qbee-agent#74](https://github.com/qbee-io/qbee-agent/pull/74)
> 
> It's still subject to testing and review, but the gist of it is that it will do a retry for the initial api call to the qbee backend on startup.

Thank you for sharing the code. When do you expect this fix to be merged?

jonhenrik13 commented:
I see your point. How about we make it configurable? There's already a string replacement going on at install time in the qbee-agent.service.in file for the path of the qbee-agent config and state files:

https://github.com/qbee-io/meta-qbee/blob/master/meta-qbee/recipes-qbee/qbee-agent/qbee-agent.inc#L65-L66

So we could simply add a new variable with the current default value, but which is override-able in local.conf? Eg. QBEE_SYSTEMD_AFTER_TARGET?

We expect the code changes related to the agent retry to be merged by the end of this week.

leon-anavi commented:
> I see your point. How about we make it configurable? There's already a string replacement going on at install time in the qbee-agent.service.in file for the path of the qbee-agent config and state files:
> 
> https://github.com/qbee-io/meta-qbee/blob/master/meta-qbee/recipes-qbee/qbee-agent/qbee-agent.inc#L65-L66
> 
> So we could simply add a new variable with the current default value, but which is override-able in local.conf? Eg. `QBEE_SYSTEMD_AFTER_TARGET`?

Yes, absolutely. Sounds as a good idea to me.

> We expect the code changes related to the agent retry to be merged by the end of this week.

OK, nice, thanks.

jonhenrik13 commented:
We're working on the 2024.45 release of the qbee-agent for Yocto currently and will try to include the configurable systemd target in this. It will hopefully be released today or tomorrow. This will also include the bump in go version.

jonhenrik13 commented:
We've just merged the recipe for the 2024.45 release of qbee-agent. It contains both the `QBEE_SYSTEMD_AFTER_TARGET` variable to configure the systemd startup targets and the bump of go version to 1.22.8. Would you be able to test the current master and see if the changes addresses the issues in your PRs?

leon-anavi commented:
> Would you be able to test the current master and see if the changes addresses the issues in your PRs?

Yes, I will test and provide feedback tomorrow morning.

leon-anavi commented:
> We've just merged the recipe for the 2024.45 release of qbee-agent. It contains both the `QBEE_SYSTEMD_AFTER_TARGET` variable to configure the systemd startup targets and the bump of go version to 1.22.8. Would you be able to test the current master and see if the changes addresses the issues in your PRs?

I tested the latest master of meta-qbee with:
```
QBEE_SYSTEMD_AFTER_TARGET="systemd-resolved.service network-online.target"
```
Unfortunately this is not enough and qbee-agent is started too early. As a follow-up from the race condition you can see the date issues in the logs:
```
root@raspberrypi4-64:~# systemctl status qbee-agent
* qbee-agent.service - The Qbee fleet management agent
     Loaded: loaded (/usr/lib/systemd/system/qbee-agent.service; enabled; preset: enabled)
     Active: active (running) since Tue 2024-02-27 17:26:09 UTC; 8 months 23 days ago
    Process: 408 ExecStartPre=/etc/qbee/yocto/qbee-bootstrap-prep.sh (code=exited, status=0/SUCCESS)
   Main PID: 429 (qbee-agent)
      Tasks: 10 (limit: 3927)
        CPU: 909ms
     CGroup: /system.slice/qbee-agent.service
             `-429 /usr/bin/qbee-agent -c /data/qbee/etc -s /data/qbee/var start

Feb 27 17:26:09 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:09 [INFO] Gathering system information
Feb 27 17:26:09 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:09 [INFO] Sending bootstrap request to device.app.qbee.io:443
Feb 27 17:26:09 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:09 [ERROR] error sending bootstrap request: failed to send API request: Post>
Feb 27 17:26:09 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:09 [INFO] Awaiting to be approved.
Feb 27 17:26:14 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:14 [INFO] Preparing agent directories
Feb 27 17:26:15 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:15 [INFO] Bootstrap successfully completed
Feb 27 17:26:15 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:15 [INFO] Please remember to start the qbee-agent service as administrative >
Feb 27 17:26:15 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:15 [INFO] Detected start command based on OS attributes is: $ systemctl star>
Feb 27 17:26:15 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:15 [INFO] Preparing agent directories
Feb 27 17:26:15 raspberrypi4-64 qbee-agent[429]: 2024/02/27 17:26:15 [INFO] starting agent scheduler
root@raspberrypi4-64:~# date
Wed Nov 20 11:35:55 UTC 2024
```
Here is the systemd service:
```
root@raspberrypi4-64:~# cat /usr/lib/systemd/system/qbee-agent.service
[Unit]
Description=The Qbee fleet management agent
After=systemd-resolved.service network-online.target

[Service]
EnvironmentFile=-/etc/default/qbee-agent
RootDirectory=/
ExecStart=/usr/bin/qbee-agent -c /data/qbee/etc -s /data/qbee/var start
ExecStartPre=/etc/qbee/yocto/qbee-bootstrap-prep.sh
ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=on-failure
#EnvironmentFile=-/etc/default/qbee-agent
# Since systemd 229, should be in [Unit] but in order to support systemd <229,
# it is also supported to have it here.
StartLimitInterval=10
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
```

I expect this is happening because the generated systemd service has `After=systemd-resolved.service network-online.target` but misses `Wants=network-online.target`.

jonhenrik13 commented:
That's odd. I might have misinterpreted the systemd docs, but I interpret them as it would be enough to set the `After` units. But I also see in other docs that the recommended way to startup a service with full network connectivity is to use both `Wants` and `After`. Anyway, I've made the following PR to also allow for overriding of the `Wants` units.

https://github.com/qbee-io/meta-qbee/pull/38

I'll let you know once it's merged.

jonhenrik13 commented:
The aforementioned PR is now merged. Would you be able to test the current master to see if that helps? You can now also define the `QBEE_SYSTEMD_WANTS_TARGET`.

leon-anavi commented:
> The aforementioned PR is now merged. Would you be able to test the current master to see if that helps? You can now also define the `QBEE_SYSTEMD_WANTS_TARGET`.

Thank you for adding it. Combined with `QBEE_SYSTEMD_AFTER_TARGET` the qbee-agent starts successfully without any warnings or errors.

jonhenrik13 commented:
Great! Are you ok with me closing this and the go1.22.8 PR?

jonhenrik13 commented:
Closing as this has been made configurable in #37 and #38
